### ISSUE TITLE
Grammars using generics' associated types generate code that fails to compile

### ISSUE THREAD

**Issue by fitzgen** [nikomatsakis/lalrpop#247]
# TLDR

Creating a grammar that is parameterized by a generic type that is bounded by a trait with associated types that are used as the return values of actions results in compilation failure of the generated code.

For example:

```rust
use std::str::FromStr;
use traits;

grammar<P>(callbacks: &mut P) where P: traits::ParseCallbacks;

pub Term: P::Term = {
    <n:Num> => n.into(),
    "(" <t:Term> ")" => t,
};

Num: P::Num = <s:r"[0-9]+"> => callbacks.number(i32::from_str(s).unwrap());
```

Where `traits::ParseCallbacks` is defined like this:

```rust
pub trait ParseCallbacks {
    type Term: From<Self::Num>;
    type Num;

    fn number(&mut self, i32) -> Self::Num;
}
```

## Steps to Reproduce

I've created a test case at https://github.com/fitzgen/lalrpop-generic-grammar-bug

```
$ git clone https://github.com/fitzgen/lalrpop-generic-grammar-bug.git`
$ cd lalrpop-generic-grammar-bug
$ cargo build
```

## Expected

Builds A-OK.

## Actual

Type errors in the generated parser code involving the grammar's generic type parameter's trait's associated types:

<details><pre><code>
$ cargo test
   Compiling lalrpop-generic-grammar v0.1.0 (file:///Users/fitzgen/src/lalrpop-generic-grammar)
error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:16:15
   |
16 |         NtNum(P::Num),
   |               ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:17:16
   |
17 |         NtTerm(P::Term),
   |                ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:18:20
   |
18 |         Nt____Term(P::Term),
   |                    ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:262:18
    |
262 |     ) -> (usize, P::Num, usize) {
    |                  ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:272:18
    |
272 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:282:18
    |
282 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:16:15
   |
16 |         NtNum(P::Num),
   |               ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:17:16
   |
17 |         NtTerm(P::Term),
   |                ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
  --> src/grammar.rs:18:20
   |
18 |         Nt____Term(P::Term),
   |                    ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:262:18
    |
262 |     ) -> (usize, P::Num, usize) {
    |                  ^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:272:18
    |
272 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error[E0433]: failed to resolve. Use of undeclared type or module `P`
   --> src/grammar.rs:282:18
    |
282 |     ) -> (usize, P::Term, usize) {
    |                  ^^^^^^^ Use of undeclared type or module `P`

error: aborting due to previous error(s)

error: Could not compile `lalrpop-generic-grammar`.
Build failed, waiting for other jobs to finish...
error: aborting due to previous error(s)

error: build failed
</code></pre></details>

The generated `__Symbol` contains associated types `P::Num` and `P::Term`, but isn't parameterized by the generic type `P` and is missing the `where P: traits::ParseCallbacks` clause as well:

https://github.com/fitzgen/lalrpop-generic-grammar-bug/blob/master/src/grammar.rs#L12-L19

```rust
pub enum __Symbol<'input> {
        Term_22_28_22(&'input str),
        Term_22_29_22(&'input str),
        Termr_23_22_5b0_2d9_5d_2b_22_23(&'input str),
        NtNum(P::Num),
        NtTerm(P::Term),
        Nt____Term(P::Term),
}
```

**Comment by fitzgen** [comment]
Or maybe I'm just not declaring that `P` needs to end up in `__Symbol` somehow? Not clear to me from reading `larlpop`'s `larlpop` grammar that something exists, but I could definitely be missing it...

**Comment by fitzgen** [comment]
I added some logging to `TypeRepr::referenced` and I think I am getting closer to the bug:

```
processing file `/Users/fitzgen/src/lalrpop-generic-grammar/src/grammar.lalrpop`
FITZGEN: calling TypeRepr::referenced
FITZGEN:     self = P::Num
FITZGEN:     returning []
FITZGEN: calling TypeRepr::referenced
FITZGEN:     self = P::Term
FITZGEN:     returning []
FITZGEN: calling TypeRepr::referenced
FITZGEN:     self = P::Term
FITZGEN:     returning []
```

If I'm understanding things correctly, I'd expect that `P` end up in the returned vector.

**Comment by fitzgen** [comment]
FWIW, turbofishing is not a workaround:

```
/Users/fitzgen/src/lalrpop-generic-grammar/src/grammar.lalrpop:6:11: 6:11 error: unexpected token: `<`

--- stderr
  pub Term: <P as traits::ParseCallbacks>::Term = {
            ^
```

**Comment by fitzgen** [comment]
Another potential workaround that didn't work:

```rust
pub type TermP<P: ParseCallbacks> = <P as ParseCallbacks>::Term;
pub type NumP<P: ParseCallbacks> = <P as ParseCallbacks>::Num;
```

and

```rust
pub Term: traits::TermP<P> = {
    <n:Num> => n.into(),
    "(" <t:Term> ")" => t,
};

Num: traits::NumP<P> = <s:r"[0-9]+"> => callbacks.number(i32::from_str(s).unwrap());
```

This results in these errors:

<details><pre><code>
$ cargo build
Compiling lalrpop-generic-grammar v0.1.0 (file:///Users/fitzgen/src/lalrpop-generic-grammar)
warning[E0122]: trait bounds are not (yet) enforced in type definitions
 --> src/traits.rs:8:1
  |
8 | pub type TermP<P: ParseCallbacks> = <P as ParseCallbacks>::Term;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning[E0122]: trait bounds are not (yet) enforced in type definitions
 --> src/traits.rs:9:1
  |
9 | pub type NumP<P: ParseCallbacks> = <P as ParseCallbacks>::Num;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied
  --> src/grammar.rs:16:15
   |
16 |         NtNum(traits::NumP<P>),
   |               ^^^^^^^^^^^^^^^^ the trait `traits::ParseCallbacks` is not implemented for `P`
   |
   = help: consider adding a `where P: traits::ParseCallbacks` bound

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied in `grammar::__parse__Term::__Symbol<'input, P>`
   --> src/grammar.rs:228:5
    |
228 | /     fn __pop_Term_22_28_22<
229 | |       'input,
230 | |       P,
231 | |     >(
...   |
237 | |         }
238 | |     }
    | |_____^ within `grammar::__parse__Term::__Symbol<'input, P>`, the trait `traits::ParseCallbacks` is not implemented for `P`
    |
    = help: consider adding a `where P: traits::ParseCallbacks` bound
    = note: required because it appears within the type `grammar::__parse__Term::__Symbol<'input, P>`
    = note: tuple elements must have `Sized` type

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied in `grammar::__parse__Term::__Symbol<'input, P>`
   --> src/grammar.rs:239:5
    |
239 | /     fn __pop_Term_22_29_22<
240 | |       'input,
241 | |       P,
242 | |     >(
...   |
248 | |         }
249 | |     }
    | |_____^ within `grammar::__parse__Term::__Symbol<'input, P>`, the trait `traits::ParseCallbacks` is not implemented for `P`
    |
    = help: consider adding a `where P: traits::ParseCallbacks` bound
    = note: required because it appears within the type `grammar::__parse__Term::__Symbol<'input, P>`
    = note: tuple elements must have `Sized` type

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied in `grammar::__parse__Term::__Symbol<'input, P>`
   --> src/grammar.rs:250:5
    |
250 | /     fn __pop_Termr_23_22_5b0_2d9_5d_2b_22_23<
251 | |       'input,
252 | |       P,
253 | |     >(
...   |
259 | |         }
260 | |     }
    | |_____^ within `grammar::__parse__Term::__Symbol<'input, P>`, the trait `traits::ParseCallbacks` is not implemented for `P`
    |
    = help: consider adding a `where P: traits::ParseCallbacks` bound
    = note: required because it appears within the type `grammar::__parse__Term::__Symbol<'input, P>`
    = note: tuple elements must have `Sized` type

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied
   --> src/grammar.rs:261:5
    |
261 | /     fn __pop_NtNum<
262 | |       'input,
263 | |       P,
264 | |     >(
...   |
270 | |         }
271 | |     }
    | |_____^ the trait `traits::ParseCallbacks` is not implemented for `P`
    |
    = help: consider adding a `where P: traits::ParseCallbacks` bound

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied
   --> src/grammar.rs:272:5
    |
272 | /     fn __pop_NtTerm<
273 | |       'input,
274 | |       P,
275 | |     >(
...   |
281 | |         }
282 | |     }
    | |_____^ the trait `traits::ParseCallbacks` is not implemented for `P`
    |
    = help: consider adding a `where P: traits::ParseCallbacks` bound

error[E0277]: the trait bound `P: traits::ParseCallbacks` is not satisfied
   --> src/grammar.rs:283:5
    |
283 | /     fn __pop_Nt____Term<
284 | |       'input,
285 | |       P,
286 | |     >(
...   |
292 | |         }
293 | |     }
    | |_____^ the trait `traits::ParseCallbacks` is not implemented for `P`
    |
    = help: consider adding a `where P: traits::ParseCallbacks` bound

error: aborting due to previous error(s)

error: Could not compile `lalrpop-generic-grammar`.

To learn more, run the command again with --verbose.
</details></pre></code>

**Comment by fitzgen** [comment]
So I started adding `TypeRef::Associated` and `TypeRepr::Associated`, filling in all the places where matches are no longer exhaustive, as well as adding `<Pupper as Doggo>::Wow` syntax to the `lrgrammar.lalrpop` -- is this the right rabbit hole to go down, @nikomatsakis ?

**Comment by nikomatsakis** [comment]
@fitzgen -- sorry for taking so long to get back to you here! So, I think it depends on your goals.

`TypeRef` is the enum that represents the syntax the user types directly. So, if you want to add support for the `<T as Trait>::Bar` form, then you probably need to modify this, as well as the parser and so forth. This seems like a secondary goal though, and probably a separate bug.

Otherwise, I think one way would be to modify just `TypeRepr`, and to modify the code that **converts** a `TypeRef` into a `type_repr` -- if memory serves, this would be the function `type_ref()` in `normalize/tyinfer/mod.rs`. You would probably want to change this code:

```rust
            TypeRef::Nominal { ref path, ref types } => {
                let types = try! {
                    types.iter().map(|t| self.type_ref(t)).collect()
                };
                Ok(TypeRepr::Nominal(NominalTypeRepr { path: path.clone(),
                                                       types: types }))
            }
```

which takes a path and have it check if the `path` begins with the name of one of the declared type parameters, in which case it might produce a distinct `TypeRepr` variant (e.g., `AssociatedType`).

I would expect `TypeRepr` to wind up with, ultimately, two new variants:

- one for explicit `<A as B>::C` syntax;
- and one for `A::T` where `A` is a generic type.

**Comment by fitzgen** [comment]
I've got something that works for my original test case but causes some compilation errors in other `larlpop-test/src/*.lalrpop` files.

To make my test case work, I had to add a bunch of checks for if the grammar has `where` clauses, and if so, copy those where clauses into various emitted functions and type definitions (`__Symbol`, `__pop_NtTerm`, `__NonTerminal`, etc..). However, because `where` clauses are opaque to us (they're just strings), we don't know if the `where` clause is necessary for a given function/type, and we end up with compilation errors because the where clause is referencing types that aren't in scope.

In order to determine if a `where` clause is required for each of these functions/types, we will need to properly parse `where` clauses into some kind of structure that we can pull all the referenced type parameters out of. This seems easy in the `T: Whatever` case, but more difficult in the case of bounds on associated types, etc. I guess *some* of this infrastructure already exists via `TypeRepr::referenced`...

**Comment by fitzgen** [comment]
Fixed tests; haven't added `for<'a`.
